jumanji/environments/logic/sliding_tile_puzzle/env.py

```diff
@@ -0,0 +1,285 @@
+        - empty_tile_position: Tuple of int32, representing the position of the empty tile.
+        - key: jax array (uint32) of shape (2,), random key used to generate random numbers
+            at each step and for auto-reset.
+    """
+
+    def __init__(
+        self,
+        generator: Optional[Generator] = None,
+        reward_fn: Optional[RewardFn] = None,
+        time_limit: int = 500,
+        viewer: Optional[Viewer[State]] = None,
+    ) -> None:
+        """Instantiate a `SlidingTilePuzzle` environment.
+
+        Args:
+            generator: callable to instantiate environment instances.
+                Defaults to `RandomGenerator` which generates puzzles with
+                a size of 5x5.
+            reward_fn: RewardFn whose `__call__` method computes the reward of an environment
+                transition. The function must compute the reward based on the current state,
+                the chosen action and the next state.
+                Implemented options are [`DenseRewardFn`, `SparseRewardFn`].
+                Defaults to `DenseRewardFn`.
+            time_limit: maximum number of steps before the episode is terminated.
+            viewer: environment viewer for rendering.
+        """
+        self.generator = generator or RandomGenerator(grid_size=5)
+        self.reward_fn = reward_fn or DenseRewardFn()
+
+        self.time_limit = time_limit
+
+        # Create viewer used for rendering
+        self._env_viewer = viewer or SlidingTilePuzzleViewer(name="SlidingTilePuzzle")
+        self.solved_puzzle = jnp.arange(self.generator.grid_size**2).reshape(
+            (self.generator.grid_size, self.generator.grid_size)
+        )
+
+    def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep[Observation]]:
+        """Resets the environment to an initial state."""
+        key, subkey = jax.random.split(key)
+        puzzle, empty_tile_position = self.generator(subkey)
+        state = State(
+            puzzle=puzzle,
+            empty_tile_position=empty_tile_position,
+            key=key,
+            step_count=0,
+        )
+        action_mask = self._get_valid_actions(empty_tile_position)
+        obs = Observation(
+            puzzle=puzzle,
+            empty_tile_position=empty_tile_position,
+            action_mask=action_mask,
+        )
+        timestep = restart(observation=obs, extras=self._get_extras(state))
+        return state, timestep
+
+    def step(
+        self, state: State, action: chex.Array
+    ) -> Tuple[State, TimeStep[Observation]]:
+        """Updates the environment state after the agent takes an action."""
+        (updated_puzzle, updated_empty_tile_position) = self._move_empty_tile(
+            state.puzzle, state.empty_tile_position, action
+        )
+        # Check if the puzzle is solved
+        done = jnp.array_equal(updated_puzzle, self.solved_puzzle)
+        horizon_reached = state.step_count >= self.time_limit
```

Comment:
This isn't used. We can either create `horizon_reached = next_state.step_count >= self.time_limit` below or delete this line.